### ISSUE TITLE
Why precision and recall are always 0?

### ISSUE THREAD

**Issue by dsyrock** [ultralytics/yolov5#1694]
## ❔Question
Why precision and recall are always 0?

## Additional context
I'm training my dataset TL.yaml, and using 400 images for training. I only need to train one class of images, and all the label files and yaml files are well prepeared. 

Then I run this command:
python e:\yolo\train.py --batch 16 --epoch 5 --data e:\yolo\data\TL.yaml --cfg e:\yolo\models\yolov5s.yaml --weights e:\yolo\weights\yolov5s.pt --single-cls --workers 0

After that, I checked the result image, the graph of precision and recall are a line of zero, like this:
[url=https://sm.ms/image/3DRWSMgkFitsyml][img]https://i.loli.net/2020/12/15/3DRWSMgkFitsyml.png[/img][/url]
[url=https://sm.ms/image/I8Yt54M2PeBd3WO][img]https://i.loli.net/2020/12/15/I8Yt54M2PeBd3WO.png[/img][/url]

So it means that is a fail training right? If so, how can I fix it? Thanks!

**Comment by NotFound1911** [comment]
> ## ❔Question
> Why precision and recall are always 0?
> 
> ## Additional context
> I'm training my dataset TL.yaml, and using 400 images for training. I only need to train one class of images, and all the label files and yaml files are well prepeared.
> 
> Then I run this command:
> python e:\yolo\train.py --batch 16 --epoch 5 --data e:\yolo\data\TL.yaml --cfg e:\yolo\models\yolov5s.yaml --weights e:\yolo\weights\yolov5s.pt --single-cls --workers 0
> 
> After that, I checked the result image, the graph of precision and recall are a line of zero, like this:
> [url=https://sm.ms/image/3DRWSMgkFitsyml][img][https://i.loli.net/2020/12/15/3DRWSMgkFitsyml.png[/img][/url]](https://i.loli.net/2020/12/15/3DRWSMgkFitsyml.png%5B/img%5D%5B/url%5D)
> [url=https://sm.ms/image/I8Yt54M2PeBd3WO][img][https://i.loli.net/2020/12/15/I8Yt54M2PeBd3WO.png[/img][/url]](https://i.loli.net/2020/12/15/I8Yt54M2PeBd3WO.png%5B/img%5D%5B/url%5D)
> 
> So it means that is a fail training right? If so, how can I fix it? Thanks!
When using the server (Ubuntu), I have the same problem as you. But I didn't have this problem on my PC (Windows)

**Comment by NotFound1911** [comment]
When using the server (Ubuntu), I have the same problem as you. But I didn't have this problem on my PC (Windows)

**Comment by HongYg** [comment]
same... on my PC (windows)

![image](https://user-images.githubusercontent.com/45381411/102185288-2c3bf000-3ef4-11eb-8bda-605d6914623c.png)

**Comment by NotFound1911** [comment]
Maybe there's something wrong with the file path.
So, **in utils/datasets.py**
`# self.img_files = list(cache.keys())  # update `
`# self.label_files = img2label_paths(cache.keys())  # update`
This may help you solve the problem.
And I made it work in this way.(Ubuntu, server)

**Comment by dsyrock** [comment]
> Maybe there's something wrong with the file path.
> So, **in utils/datasets.py**
> `# self.img_files = list(cache.keys()) # update `
> `# self.label_files = img2label_paths(cache.keys()) # update`
> This may help you solve the problem.
> And I made it work in this way.(Ubuntu, server)

Thanks. I edited it as you say, but I still got the zero result.

**Comment by dsyrock** [comment]
Seems that because the epoch is not enough. Sorry, my bad.

**Comment by glenn-jocher** [comment]
@HongYg you have no labels in your test set. No labels means no boxes or classes to create loss.

**Comment by sbbug** [comment]
me too ,hello, have you solved it?

**Comment by HongYg** [comment]
> @HongYg 테스트 세트에 레이블이 없습니다. 레이블이 없다는 것은 손실을 초래할 상자 나 클래스가 없음을 의미합니다.
I solved it! thank you.
my problem was test dataset!

**Comment by AlexJunCpp** [comment]
me too, do you know how to fix it ?

**Comment by shubhambagwari** [comment]
I forked the latest version of Yv5, Still, I am facing this error.
Earlier I trained this model with 2000 images but faced the same issue.

Roboflow has been used to label the custom dataset.
I have read the earlier discussions but not working.



![207236797-e2cf2e4d-67ec-4a52-8fdb-ff048cf0caa3](https://user-images.githubusercontent.com/25363312/207513773-032f1cad-25d7-4c58-ae86-13d9b9006afb.png)

**Comment by glenn-jocher** [comment]
@shubhambagwari train 300 epochs.

**Comment by manicmojo** [comment]
I am getting same issue. I have trained over 30 successful models. 
This model is built off same data as a successfully trained model built on yolov8s.pt. All 30 have trained on yolov8.0.181. I see there is an update, but want to fix this error before I update.

The previous models were on the same CLI, and same data.yaml, THOUGH, I have used a different format this time in data.yaml but should be fine. This format:
`train:
  - /media/xxxx/xxx/xxxx/data/xxx/xxx.train
  
val: 
  - /media/xxxx/xxx/xxxx/data/xxx/xxx.valid

nc: 32
names: ['.......]`

This is the CLI for the 0 precision results. 

`yolo detect train data=/home/xxxxx/ultralytics/ultralytics/yolo/data/data.yaml model=/home/xxxxx/ultralytics/yolov8n.pt epochs=100 imgsz=2560 device=0 name=xxxxxx batch=8 dynamic=true
New https://pypi.org/project/ultralytics/8.1.42 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.181 🚀 Python-3.8.10 torch-2.0.0+nv23.05 CUDA:0 (Orin, 62796MiB)
engine/trainer: task=detect, mode=train, model=/home/xxxx/ultralytics/yolov8n.pt, data=/home/xxx/ultralytics/ultralytics/yolo/data/data.yaml, epochs=100, patience=50, batch=8, imgsz=2560, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=xxxxxxx, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, stream_buffer=False, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=True, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=/home/xxxx/ultralytics/runs/xxxx/xxxxxx
Overriding model.yaml nc=80 with nc=32
`

I have ran the validator on this model again with the same result:

`Ultralytics YOLOv8.0.181 🚀 Python-3.8.10 torch-2.0.0+nv23.05 CUDA:0 (Orin, 62796MiB)
Model summary (fused): 168 layers, 3011888 parameters, 0 gradients
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%
                   all       3437      35521          0          0          0          0
Speed: 1.4ms preprocess, 30.0ms inference, 0.0ms loss, 0.4ms postprocess per image
`

And repeated it on the previous yolov8s.pt model, which gave me normal results..
20k images, only key difference is yolov8n.pt size...

Thoughts? Thanks.

**Comment by glenn-jocher** [comment]
@manicmojo it sounds like there might be an issue with either the format of your `data.yaml` or potentially a compatibility concern with the new model architecture `yolov8n.pt` and your setup. Since your previous models and data worked successfully, let's try to isolate the problem:

1. **Check the `data.yaml` format**: Ensure the paths are correct and the format is accurately followed as per the YOLOv5 documentation. There might be subtle differences that are causing issues with detection.

2. **Test with a Smaller Dataset**: Try training with a smaller subset of your dataset that previously worked with `yolov8s.pt`. This can help identify if the issue is dataset-specific or model-specific.

3. **Validate Data and Annotations**: Double-check that your dataset and annotations are correctly formatted and there are no missing labels or mismatches that could cause the model to fail at detecting objects.

4. **Update to the Latest Version**: While you mentioned wanting to fix this error before updating, it might be beneficial to update to the latest version since there could be bug fixes or improvements that resolve your issue.

5. **Revert to Previous Working Conditions**: If none of the above steps work, try training again under the exact conditions (data, CLI arguments, etc.) when it was last successful to further isolate the issue.

If after trying these steps you still encounter issues, consider providing more detailed information on the changes in your `data.yaml` and any additional logs for further diagnosis. Remember, keeping all components up-to-date and ensuring dataset integrity are critical for successful model training.
